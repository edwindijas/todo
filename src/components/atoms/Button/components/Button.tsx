interface ButtonProps {
  children: React.ReactElement | string,
  className: string,
  onClick?: () => void,
  dataTestId: string
}

 const Button = ({
  className,
  children,
  onClick,
  dataTestId
}: ButtonProps): JSX.Element =>
  <button data-testid={dataTestId} onClick={onClick} className={className}>
    {children}
  </button>


export default Button