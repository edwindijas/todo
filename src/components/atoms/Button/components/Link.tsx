import {useCallback} from 'react'
import { cva } from 'class-variance-authority'
import { Link, NavLink } from 'react-router-dom'

interface CustomLinkProps {
  isNavLink: boolean,
  children: React.ReactElement | string ,
  className: string,
  to: string,
  onClick?: () => void
}

const customLinkVariants = cva('', {
  variants: {
    isActive: {
      true: ['bg-blue-light text-blue-medium fill-blue-medium relative',
      'before:absolute before:top-2 before:bottom-2 before:left-0 before:bg-blue-medium before:w-1 before:block before:rounded-sm'
      ],
      false: 'text-blue-desaturated fill-blue-desaturated'
    }
  },
  defaultVariants: {
    isActive: false
  }
})

const CustomLink = ({
  children,
  className,
  isNavLink,
  to,
  onClick
}: CustomLinkProps): JSX.Element => {

  const getClassName = useCallback(({isActive}: {isActive: boolean}): string => {
    const cvaClass = customLinkVariants({isActive})
    return `${className} ${cvaClass}`
  }, [className])

  if (isNavLink) {
    return<NavLink to={to} className={getClassName} >{children}</NavLink>
  }

  return <Link to={to} className={className} onClick={onClick} >{children}</Link>
}




export default CustomLink 