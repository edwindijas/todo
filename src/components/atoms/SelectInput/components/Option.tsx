import { cva } from 'class-variance-authority'
import { useMemo } from 'react'

const SelectOption = ({item, selected}: SelectOptionProps): JSX.Element => {

  const isSelected = useMemo<boolean>((): boolean => {

  if (typeof item === 'string' && typeof selected === 'string') {
    return item === selected
  }
  
    return false
  }, [selected])

  const label = typeof item === 'string' ? item : item.value
  return <div className={selectOptionVariants({selected: isSelected})} >
    { label }
  </div>
}


const selectOptionVariants = cva('', {
  variants: {
    selected: {
      true: 'bg-blue-medium text-white'
    }
  }
})


interface SelectOptionProps {
  item: SelectOptionItem
  selected: SelectOptionItem
  onClick?: (v: SelectOptionItem) => void
}

export type SelectOptionItem = string | {
  value: string
  id: string
}

export default SelectOption