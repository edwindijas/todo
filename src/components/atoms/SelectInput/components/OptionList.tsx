import { useState } from 'react'
import SelectOption, { type SelectOptionItem } from './Option'

const OptionsList = (
  {
    options
  }: OptionListProps
): JSX.Element => {
  const [selected, setSelected] = useState<SelectOptionItem>('')
  
  return  <ul >
  {
    options.map((option) => <li role='select-option' key={typeof option === 'string' ? option : option.id} >
      <SelectOption onClick={setSelected} selected={selected} item={option} />
    </li>)
  }
</ul>
}

interface OptionListProps {
  options: SelectOptionItem[]
}

export default OptionsList