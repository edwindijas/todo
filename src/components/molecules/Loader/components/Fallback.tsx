const Fallback = (): JSX.Element => (
  <div className='w-full h-full flex items-center justify-center'>
    Loading...
  </div>
)
export default Fallback