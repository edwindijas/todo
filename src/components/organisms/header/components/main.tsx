import LogoIcon from '@/components/atoms/Icon/Logo'
import Menu, { type MenuItem } from '@/components/molecules/menu'

const menuItems: MenuItem[] = [
  {
    label: 'ToDo',
    to: '/todo'
  },
  {
    label: 'Calendar',
    to: '/calendar'
  },
  {
    label: 'Board',
    to: '/board'
  }
]

const Main = (): JSX.Element => <div className='w-64 border-r border-solid h-full'>
  <div className='flex p-4 gap-2 items-center'>
    <div className='w-10 h-10'><LogoIcon /></div>
    <p className='text-lg flex flex-col gap-0 leading-6'>
      <span>Task</span>
      <span className='font-bold'>Flow</span>
    </p>
  </div>
  <div>

  </div>
  <div className='p-2' >
    <Menu items={menuItems} isNavLink />
  </div>
</div>

export default Main