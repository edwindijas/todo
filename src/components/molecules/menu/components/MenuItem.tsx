import Button from '@/components/atoms/Button'

export interface MenuItemProps {
  startIcon?: React.ReactElement,
  endIcon?: React.ReactElement,
  label: string,
  to: string
  isNavLink?: boolean
}

const MenuItem = ({
  startIcon,
  endIcon,
  label,
  to,
  isNavLink,
}: MenuItemProps): JSX.Element =>
 <Button
  startIcon={startIcon}
  isNavLink={isNavLink}
  isLink to={to}
  endIcon={endIcon}
  padding='lg'
>{ label }</Button>

export default MenuItem
