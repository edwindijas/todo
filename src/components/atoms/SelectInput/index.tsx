import { useRef } from 'react'
import { type SelectOptionItem } from './components/Option'
import OptionsList from './components/OptionList'
import Button from '../Button'
import ExpandCircleDownIcon from '../Icon/ExpandCircleDown'
import useExpandable from '@/hooks/useExpandable'


const SelectInput = ({
  label,
  options = []
}: SelectProps): JSX.Element => {
  const wrapper = useRef<HTMLDivElement | null>(null)
  const [active, toggleDropDown] = useExpandable({wrapper})

  return <div role='select' ref={wrapper} >
    <Button
      dataTestId='btn-toggle-select'
      intent='secondary'
      onClick={toggleDropDown} endIcon={<ExpandCircleDownIcon />}><p>{label}</p></Button>
    { 
      active && 
      <OptionsList options={options} />
    }
  </div>
}

interface SelectProps  {
  defaultValue?: string
  label?: string,
  options?: SelectOptionItem[]
}

export default SelectInput