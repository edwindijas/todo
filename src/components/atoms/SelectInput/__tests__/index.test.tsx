import { fireEvent } from '@testing-library/react'
import render from '@/utils/test-utils/customRender'
import SelectInput from '..'
import { type SelectOptionItem } from '../components/Option'

const menu: SelectOptionItem[] = [
  'Wowo',
  {
    value: 'Hey',
    id: 'people'
  }
]

describe('Select Input', () => {
  it('Matches snapshot', () => {
    const {container, getAllByTestId} = render(<>
      <SelectInput label="Hello World" options={menu} />
      <SelectInput label="Hello World" />
    </>)
    fireEvent.click(getAllByTestId('btn-toggle-select')[0])
    expect(container).toMatchSnapshot()
  })
})