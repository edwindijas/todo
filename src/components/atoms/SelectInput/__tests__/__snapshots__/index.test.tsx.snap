// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Input Matches snapshot 1`] = `
<div>
  <div
    role="select"
  >
    <button
      class=" flex w-full capitalize items-center gap-2 relative text-center border border-blue-medium bg-blue-light text-blue-medium fill-blue-medium px-2 h-14 rounded-md"
      data-testid="btn-toggle-select"
    >
      <span
        class="w-full"
      >
        <p>
          Hello World
        </p>
      </span>
      <div>
        <svg
          class=" w-full h-full"
          data-testid="component-atom-icon"
        />
      </div>
    </button>
    <ul>
      <li
        role="select-option"
      >
        <div
          class=""
        >
          Wowo
        </div>
      </li>
      <li
        role="select-option"
      >
        <div
          class=""
        >
          Hey
        </div>
      </li>
    </ul>
  </div>
  <div
    role="select"
  >
    <button
      class=" flex w-full capitalize items-center gap-2 relative text-center border border-blue-medium bg-blue-light text-blue-medium fill-blue-medium px-2 h-14 rounded-md"
      data-testid="btn-toggle-select"
    >
      <span
        class="w-full"
      >
        <p>
          Hello World
        </p>
      </span>
      <div>
        <svg
          class=" w-full h-full"
          data-testid="component-atom-icon"
        />
      </div>
    </button>
  </div>
</div>
`;
